extensions/amp-subscriptions-google/0.1/amp-subscriptions-google.js

```diff
@@ -571,70 +617,205 @@ export class GoogleSubscriptionsPlatform {
-    // whether the access has been granted and whether user is a subscriber.
-    if (!best.granted) {
-      this.runtime_.showOffers({list: 'amp'});
-    } else if (!best.isSubscriber()) {
-      this.runtime_.showAbbrvOffer({list: 'amp'});
-    }
+
+    const googleMeteringStrategyPromise = this.getGoogleMeteringStrategy_();
+    const meteringStatePromise = this.serviceAdapter_.loadMeteringState();
+    const promises = Promise.all([
+      googleMeteringStrategyPromise,
+      meteringStatePromise,
+    ]);
+
+    promises.then((results) => {
+      const meteringStrategy = results[0];
+      const meteringState = results[1];
+
+      if (meteringStrategy === GoogleMeteringStrategy.EXTENDED_ACCESS) {
+        // Show the Regwall, so the user can get
+        // a metering state that leads to a
+        // granting entitlement.
+        // After the Regwall flow completes, then continue authorization flow.
+        if (!best.granted && !meteringState) {
+          this.showMeteringRegwall_().then(continueAuthorizationFlow);
+          return;
+        }
+
+        // Consume the metering entitlement.
+        if (best.granted && !best.isSubscriber()) {
+          this.runtime_.consumeShowcaseEntitlementJwt(
+            best.raw,
+            continueAuthorizationFlow
+          );
+          return;
+        }
+      }
+
+      // Offers or abbreviated offers may need to be shown depending on
+      // whether the access has been granted and whether user is a subscriber.
+      if (!best.granted) {
+        this.runtime_.showOffers({list: 'amp'});
+      } else if (!best.isSubscriber()) {
+        this.runtime_.showAbbrvOffer({list: 'amp'});
+      }
+    });
+  }
+
+  /**
+   * Asks user to register an account with the publisher.
+   * Returns a promise that resolves when the process completes.
+   * @return {!Promise}
+   */
+  showMeteringRegwall_() {
+    // Show the Showcase Regwall.
+    const googleSignInDetailsPromise = GaaMeteringRegwall.show({
+      // Specify a URL that renders a Google Sign-In button.
+      iframeUrl: this.serviceConfig_['googleSignInHelperUrl'],
+    });
+    const ampReaderIdPromise = this.serviceAdapter_.getReaderId('local');
+
+    // Register the user with the publisher.
+    const registerUserPromise = Promise.all([
+      googleSignInDetailsPromise,
+      ampReaderIdPromise,
+    ]).then((results) => {
+      const googleSignInDetails = results[0];
+      const ampReaderId = results[1];
+
+      const url = this.serviceConfig_['meteringRegistrationUrl'];
+      const postBody = {
+        googleSignInDetails,
+        ampReaderId,
+      };
+
+      return this.fetcher_.sendPostToPublisher(url, postBody);
+    });
+
+    // The publisher responds with a metering state.
+    // Let's save it.
+    const saveMeteringStatePromise = registerUserPromise.then(
+      (publisherResponse) => {
+        const meteringState =
+          publisherResponse &&
+          publisherResponse['metering'] &&
```

Comment:
I think expecting a single `userState` object in the response from the publisher would be better. We can then transform into metering.state if needed.